snap/naming/validate.go

```diff
@@ -174,3 +175,45 @@ func ValidateSnapID(id string) error {
 	}
 	return nil
 }
+
+var errInvalidSecurityTag = errors.New("invalid security tag")
+
+// ValidateSecurityTag validates known variants of snap security tag.
+//
+// Two forms are recognised, one for apps and one for hooks. Other forms
+// are possible but are not handled here.
+//
+// TODO: handle the weird udev variant.
+func ValidateSecurityTag(tag string) error {
+	parts := strings.Split(tag, ".")
+	switch len(parts) {
+	case 3:
+		snapLiteral, snapName, appName := parts[0], parts[1], parts[2]
+		if snapLiteral != "snap" {
+			return errInvalidSecurityTag
+		}
+		if err := ValidateInstance(snapName); err != nil {
+			return errInvalidSecurityTag
+		}
+		if err := ValidateApp(appName); err != nil {
+			return errInvalidSecurityTag
+		}
+		return nil
+	case 4:
+		snapLiteral, snapName, hookLiteral, hookName := parts[0], parts[1], parts[2], parts[3]
+		if snapLiteral != "snap" {
+			return errInvalidSecurityTag
+		}
+		if err := ValidateInstance(snapName); err != nil {
+			return errInvalidSecurityTag
+		}
+		if hookLiteral != "hook" {
```

Comment:
I would replace the order of this check with the `ValidateInstance(..)` one. Let the constants be checked first before we start validating the snap & hook names.

Reply:
Done